src/thread.rs

```diff
@@ -0,0 +1,583 @@
+    /// threads).
+    Terminated,
+}
+
+/// The join status of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ThreadJoinStatus {
+    /// The thread can be joined.
+    Joinable,
+    /// A thread is detached if its join handle was destroyed and no other
+    /// thread can join it.
+    Detached,
+    /// The thread was already joined by some thread and cannot be joined again.
+    Joined,
+}
+
+/// A thread.
+pub struct Thread<'mir, 'tcx> {
+    state: ThreadState,
+    /// Name of the thread.
+    thread_name: Option<Vec<u8>>,
+    /// The virtual call stack.
+    stack: Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>>,
+    /// The join status.
+    join_status: ThreadJoinStatus,
+}
+
+impl<'mir, 'tcx> Thread<'mir, 'tcx> {
+    /// Check if the thread terminated. If yes, change the state to terminated
+    /// and return `true`.
+    fn check_terminated(&mut self) -> bool {
+        if self.state == ThreadState::Enabled {
+            if self.stack.is_empty() {
+                self.state = ThreadState::Terminated;
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl<'mir, 'tcx> std::fmt::Debug for Thread<'mir, 'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(ref name) = self.thread_name {
+            if let Ok(name_str) = std::str::from_utf8(name) {
+                write!(f, "{}", name_str)?;
+            } else {
+                write!(f, "<invalid thread name>")?;
+            }
+        } else {
+            write!(f, "unnamed")?;
+        }
+        write!(f, "({:?}, {:?})", self.state, self.join_status)
+    }
+}
+
+impl<'mir, 'tcx> Default for Thread<'mir, 'tcx> {
+    fn default() -> Self {
+        Self {
+            state: ThreadState::Enabled,
+            thread_name: None,
+            stack: Vec::new(),
+            join_status: ThreadJoinStatus::Joinable,
+        }
+    }
+}
+
+/// A set of threads.
+#[derive(Debug)]
+pub struct ThreadManager<'mir, 'tcx> {
+    /// Identifier of the currently active thread.
+    active_thread: ThreadId,
+    /// Threads used in the program.
+    ///
+    /// Note that this vector also contains terminated threads.
+    threads: IndexVec<ThreadId, Thread<'mir, 'tcx>>,
+    /// A counter used to generate unique identifiers for blocksets.
+    blockset_counter: u32,
+    /// A mapping from a thread-local static to an allocation id of a thread
+    /// specific allocation.
+    thread_local_alloc_ids: RefCell<FxHashMap<(DefId, ThreadId), AllocId>>,
+    /// A flag that indicates that we should change the active thread.
+    yield_active_thread: bool,
+}
+
+impl<'mir, 'tcx> Default for ThreadManager<'mir, 'tcx> {
+    fn default() -> Self {
+        let mut threads = IndexVec::new();
+        // Create the main thread and add it to the list of threads.
+        let mut main_thread = Thread::default();
+        main_thread.join_status = ThreadJoinStatus::Detached;
```

Comment:
```suggestion
        // The main thread can *not* be joined on.
        main_thread.join_status = ThreadJoinStatus::Detached;
```
Do we have a test for this?